### PR TITLE
GAX additions to support bidirectional streaming.

### PR DESCRIPTION
**Pull request by chrisdunelm**

### CODE REVIEW

File: src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Completes the stream when all buffered messages have been sent. This method can only be called
+        /// once, and further messages cannot be written after it has been called.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">This method has already been called.</exception>
+        /// <returns>A <see cref="Task"/> which will complete when the stream has finished being completed.</returns>
+        public virtual Task WriteCompleteAsync()
```
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs]: Not a full review, but two points before I forget:

- I suspect we also want TryWriteCompleteAsync, having now used this. (Could manage without, but can be handy sometimes.)
- We should rename CompleteAsync to WriteCompleteAsync in the BufferedClientStreamWriter.

Comment (chrisdunelm) [inline src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs]: Done

File: src/Google.Api.Gax.Grpc/BidirectionalStreamingSettings.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        /// </summary>
+        /// <param name="bufferCapacity">The capacity of the write buffer.</param>
+        /// <returns>A <see cref="BidirectionalStreamingSettings"/> configured with a write buffer capacity.</returns>
+        public static BidirectionalStreamingSettings FromBufferCapacity(int bufferCapacity) =>
+            new BidirectionalStreamingSettings(bufferCapacity);
+
+        /// <summary>
+        /// Configure settings for bidirection streaming.
+        /// </summary>
+        /// <param name="bufferedClientWriterCapacity">The capacity of the write buffer.</param>
```
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/BidirectionalStreamingSettings.cs]: Do we need a public constructor and a factory method?

(bidirection => bidirectional too, btw)

File: src/Google.Api.Gax.Grpc/ClientHelper.cs [hunk 3]
```diff
@@ -69,5 +70,27 @@ public ClientHelper(ServiceSettingsBase settings)
...
+        /// <param name="streamingSettings">The default streaming settings.</param>
+        /// <returns>An API call to proxy to the RPC calls</returns>
+        public ApiBidirectionalStreamingCall<TRequest, TResponse> BuildApiCall<TRequest, TResponse>(
+            Func<CallOptions, AsyncDuplexStreamingCall<TRequest, TResponse>> grpcCall,
+            CallSettings perMethodCallSettings,
+            BidirectionalStreamingSettings streamingSettings)
+            where TRequest : class, IMessage<TRequest>
+            where TResponse : class, IMessage<TResponse>
+        {
+            // TODO: What about retry?
```
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/ClientHelper.cs]: Looks like retry is basically not an option - so maybe we should check that there *is* no retry?

File: src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs [hunk 4]
```diff
@@ -88,9 +88,9 @@ public virtual Task WriteAsync(TRequest message, WriteOptions options)
         /// Completes the stream when all buffered messages have been sent. This method can only be called
         /// once, and further messages cannot be written after it has been called.
```
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs]: TryWriteComplete can be called multiple times, but only the first time will have any effect (at the moment - that will change when we put it in the buffer, of course).

File: src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs [hunk 5]
```diff
@@ -85,8 +85,9 @@ public virtual Task WriteAsync(TRequest message, WriteOptions options)
         }
 
         /// <summary>
-        /// Completes the stream when all buffered messages have been sent. This method can only be called
-        /// once, and further messages cannot be written after it has been called.
+        /// Completes the stream when all buffered messages have been sent.
+        /// Only the first call to this method on any instance will have any affect;
```
Comment (jskeet) [inline src/Google.Api.Gax.Grpc/BidirectionalStreamingBase.cs]: Nit: affect => effect :)

**Pull request merged**